pkg/querier/queryrange/codec.go

```diff
@@ -744,23 +745,72 @@ func (Codec) MergeResponse(responses ...queryrangebase.Response) (queryrangebase
 		}, nil
 	case *VolumeResponse:
 		resp0 := responses[0].(*VolumeResponse)
-		headers := resp0.Headers
+		headers := make([]*definitions.PrometheusResponseHeader, len(resp0.Headers))
+		for i, header := range resp0.Headers {
+			h := header
+			headers[i] = &h
+		}
 
 		resps := make([]*logproto.VolumeResponse, 0, len(responses))
 		for _, r := range responses {
 			resps = append(resps, r.(*VolumeResponse).Response)
 		}
 
-		return &VolumeResponse{
-			Response: seriesvolume.Merge(resps, resp0.Response.Limit),
-			Headers:  headers,
-		}, nil
+		promResponse := queryrangebase.PrometheusResponse{
+			Status:  loghttp.QueryStatusSuccess,
+			Data:    MergeToPrometheusResponse(resps, resp0.Response.Limit),
+			Headers: headers,
+		}
 
+		return &LokiPromResponse{
+			Response:   &promResponse,
+			Statistics: stats.Result{},
+		}, nil
 	default:
 		return nil, errors.New("unknown response in merging responses")
 	}
 }
 
+func MergeToPrometheusResponse(responses []*logproto.VolumeResponse, limit int32) queryrangebase.PrometheusData {
+	mergedVolumes := seriesvolume.Merge(responses, limit)
+	return mapToPrometheusResponse(mergedVolumes)
+}
+
+func mapToPrometheusResponse(mergedResponse *logproto.VolumeResponse) queryrangebase.PrometheusData {
+	samplesByStream := map[string]*logproto.LegacySample{}
+
+	// since this is an instant response, we're only interested in a single "bucket", which is
+	// bounded by the latest timestamp we've seen (which should correspond to the end of the query range)
+	tsMs := mergedResponse.Through.UnixNano() / 1e6 //convert ns to ms
+
+	// Aggregate samples into single sample with latest timestamp
+	for _, volume := range mergedResponse.Volumes {
+		if sample, ok := samplesByStream[volume.Name]; !ok {
```

Comment:
Extreme nit: omit the else statement:

```go
for _, volume := range mergedResponse.Volumes {
  if _, ok := samplesByStream[volume.Name]; !ok {
	  samplesByStream[volume.Name] = &logproto.LegacySample{
		  TimestampMs: tsMs,
	  }
  }
  samplesByStream[volume.Name].Value += float64(volume.Volume)
}
```

Reply:
The `0` value we finally click with this old Java developer some day...